components/forms/src/api.rs

```diff
@@ -0,0 +1,553 @@
+    // in an i64 (a date 1000 years in the future, for example), but
+    // appropriately handling that is complex.
+    Ok(i64::deserialize(deserializer).unwrap_or_default().max(0))
+}
+
+#[allow(dead_code)]
+pub fn add_address(conn: &mut Connection, new_address: NewAddressFields) -> Result<Address> {
+    let tx = conn.transaction()?;
+    let now_ms = get_current_time_ms();
+
+    let address = Address {
+        guid: Guid::random(),
+        address_fields: new_address,
+        time_created: now_ms,
+        time_last_used: 0,
+        time_last_modified: now_ms,
+        times_used: 0,
+        sync_change_counter: 1,
+    };
+
+    tx.execute_named(
+        &format!(
+            "INSERT OR IGNORE INTO addresses_data (
+                {common_cols},
+                sync_change_counter
+            ) VALUES (
+                :guid,
+                :given_name,
+                :additional_name,
+                :family_name,
+                :organization,
+                :street_address,
+                :address_level3,
+                :address_level2,
+                :address_level1,
+                :postal_code,
+                :country,
+                :tel,
+                :email,
+                :time_created,
+                :time_last_used,
+                :time_last_modified,
+                :times_used,
+                {sync_change_counter}
+            )",
+            common_cols = COMMON_COLS,
+            sync_change_counter = address.sync_change_counter
+        ),
+        rusqlite::named_params! {
+            ":guid": address.guid,
+            ":given_name": address.address_fields.given_name,
+            ":additional_name": address.address_fields.additional_name,
+            ":family_name": address.address_fields.family_name,
+            ":organization": address.address_fields.organization,
+            ":street_address": address.address_fields.street_address,
+            ":address_level3": address.address_fields.address_level3,
+            ":address_level2": address.address_fields.address_level2,
+            ":address_level1": address.address_fields.address_level1,
+            ":postal_code": address.address_fields.postal_code,
+            ":country": address.address_fields.country,
+            ":tel": address.address_fields.tel,
+            ":email": address.address_fields.email,
+            ":time_created": address.time_created,
+            ":time_last_used": address.time_last_used,
+            ":time_last_modified": address.time_last_modified,
+            ":times_used": address.times_used,
+        },
+    )?;
+
+    tx.commit()?;
+    Ok(address)
+}
+
+#[allow(dead_code)]
+pub fn get_address(conn: &mut Connection, guid: &Guid) -> Result<Address> {
+    let tx = conn.transaction()?;
+    let sql = format!(
+        "
+        SELECT
+            {common_cols},
+            sync_change_counter
+        FROM addresses_data d
+        WHERE guid = :guid
+            AND NOT EXISTS (
```

Comment:
If `delete_address` deletes the row from `addresses_data` (and we rely on a trigger or something to make sure a GUID can't be in both `addresses_data` and `addresses_tombstones`), we can avoid the `NOT EXISTS` checks 😄